proposals/NNNN-unlock-existential-types-for-all-protocols.md

```diff
@@ -0,0 +1,345 @@
+* This approach seems likely to lead to trade-offs between optimal design and compliance with the modifier.
+* Being a source-compatible addition, modifiers can merely offer the *option* to be explicit.
+
+### Syntactic Matters First
+
+So far, existentials are the only built-in abstraction model (on par with generics and opaque types) that doesn't have its own idiosyncratic syntax; they are spelled as the bare protocol name or a composition thereof. Although the syntax strongly suggets that the protocol as a type and the protocol as a constraint are one thing, in practice, they serve different purposes, and this manifests most confusingly in the *"Protocol (the type) does not conform to Protocol (the constraint)"* paradox. This could be qualified as a missing feature in the language; the bottom line is that the syntax is tempting developers when it should be contributing to weighted decisions. Because using existential types is syntactically lightweight in comparison to using other abstractions, and similar to using a base class, the more possibilities they offer, the more *users are* vulnerable to unintended or inappropriate type erasure by following the path of initial least resistance.
+
+With regard to a source-compatible adoption of new adornments, we could
+* force all newly supported existential types to use the new syntax, i.e. `Any<P>` or [`any P` as a dual to `some P`](https://forums.swift.org/t/improving-the-ui-of-generics/22814#heading--clarifying-existentials), and
+* allow the respective protocols of all newly supported existential types to be annotated with an attribute to opt into the old syntax (to allow for source-compatible addition of requirements).
+
+Though, if we *were* going to introduce a new syntax for existentials, we think it'd be much less confusing if we took the potentially source-breaking path and did so uniformly, deprecating the existing syntax after a late-enough language version, than to have yet another attribute and two syntaxes where one only works some of the time. We also believe that drawing a tangible line between protocols that "do" and "do not" have limited access to their API is ill-advised due to the relative nature of this phenomenon.
+
+### Leave the Language in Its Existing State
+
+The current semantics regarding existential types prevent language users, especially library authors, from inheriting useful protocols due to the high cost of maintaining a type erasure construct. Namely, the high maintenance cost is a result of type erasure containers constituting hard-to-understand boilerplate code that entails thorough testing and a significant effort when expanding the protocol at hand. Furthermore, regular language users are often confused by the current restriction, and often lack the expertise or resources required for creating correct type-erasure contructs. 
+
+## Future Directions
```

Comment:
Do you think that we should add a future direction for opening an existential, and then wrapping it into a type-erasure container? For instance, `init(_ box: Hashable)` could be added to `AnyHashable`. I'm not sure, though, if this addition could be briefly stated in an existing section, or if we'd have to create a new one.

Reply:
There's a bullet point for "automatically opening existentials when passed as generic arguments" at the bottom. That would allow the generic initializer of `AnyHashable` to accept a `Hashable` value.

Reply:
That would require some implementation work; at the same time, users will be left unable to initialize standard-library-provided type erasers with their language-provided protocol counterparts. I'm not sure if this feature should be part of the proposal, because I think it could create a major usability issue, due to the lack of a feature that can be easily added to the standard library. 

Sorry, if I've missed some aspect of the proposed feature that already enables this behavior.

Reply:
Sorry, the part about "automatically opening existential values" is a future direction (it was also mentioned as such a couple of times in the forums); we are not proposing or implementing it now.

> users will be left unable to initialize standard-library-provided type erasers with their language-provided protocol counterparts.

In case this is still relevant — why is that? Self-conformance *does* allow you to initialize an `AnyProtocol` wrapper with a `Protocol` value, among other stuff.

Reply:
I wasn't very clear. 

I'm arguing that implementing "self-conforming" existential types would require some implementation work, which is — obviously — not in scope for this proposal.

Yet, the current proposal (without self-conformance) creates serious usability issues — provided that I haven't missed anything — since users will be unable to write something like `AnyHashable(1 as Hashable)`.

What I propose we do, is include:

```swift
extension AnyHashable {
  public init(_ box: Hashable) {
    self = _openExistential(box, do: { unboxed in
      AnyHashable(unboxed)
    })
  }
}
```

Similarly, we could provide such an initializer for `AnyCollection` and any other type-erasing, standard-library wrappers.

Reply:
I see. But `AnyCollection` is not really an option, because we need to somehow represent the `Element` type once we get the wrapper back. 

I have mixed feelings about a permanent workaround to a temporary issue, given that users can add the workaround themselves if they happen to need a new `Hashable` -> `AnyHashable` conversion in the meantime. You are totally welcome to add it as a future direction though.

Reply:
> I see. But AnyCollection is not really an option, because we need to somehow represent the Element type once we get the wrapper back.

Right. This, then, leaves us just with `AnyHashable`.

> I have mixed feelings about a permanent workaround to a temporary issue, given that users can add the workaround themselves if they happen to need a new Hashable -> AnyHashable conversion in the meantime. You are totally welcome to add it as a future direction though.

I understand your concern. I think whether a workaround will be implemented depends on the timing of the follow-up proposal for automatically opening existentials.

How about the following for a future direction bullet: 

> Add an `init(_ box: Hashable)` initializer to `AnyHashable` to alleviate confusion and aid in usability. This initializer should be treated as a workaround, and removed once automatic opening of existential types becomes available.

Reply:
Sounds fine, but

> ...and removed once automatic opening of existential types becomes available.

this would break the ABI (we cannot remove something `public` once added).

Reply:
Yeah, I meant "deprecated". Then, after a major version of being deprecated, it could be removed — if that's an option. That, of course, entails that `init(_ box: Hashable)` be marked `@_disfavoredOverload`, in addition to deprecated.

Is this okay?
> Add an `init(_ box: Hashable)` initializer to `AnyHashable` to alleviate confusion and aid in usability. This initializer should be treated as a workaround, and **_deprecated_** once automatic opening of existential types becomes available.

Reply:
* should -> would
* once -> if/when

LGTM.

Reply:
I just changed `if/when ... becomes available` to `should ... become available`.